src/ir/bits.h

```diff
@@ -128,35 +128,98 @@ struct DummyLocalInfoProvider {
 template<typename LocalInfoProvider = DummyLocalInfoProvider>
 Index getMaxBits(Expression* curr,
                  LocalInfoProvider* localInfoProvider = nullptr) {
-  if (auto* const_ = curr->dynCast<Const>()) {
+  if (auto* c = curr->dynCast<Const>()) {
     switch (curr->type.getSingle()) {
       case Type::i32:
-        return 32 - const_->value.countLeadingZeroes().geti32();
+        return 32 - c->value.countLeadingZeroes().geti32();
       case Type::i64:
-        return 64 - const_->value.countLeadingZeroes().geti64();
+        return 64 - c->value.countLeadingZeroes().geti64();
       default:
         WASM_UNREACHABLE("invalid type");
     }
   } else if (auto* binary = curr->dynCast<Binary>()) {
     switch (binary->op) {
       // 32-bit
-      case AddInt32:
-      case SubInt32:
-      case MulInt32:
-      case DivSInt32:
-      case DivUInt32:
-      case RemSInt32:
-      case RemUInt32:
       case RotLInt32:
       case RotRInt32:
+      case SubInt32:
+        return 32;
+      case AddInt32: {
+        auto maxBitsLeft = getMaxBits(binary->left, localInfoProvider);
+        auto maxBitsRight = getMaxBits(binary->right, localInfoProvider);
+        return std::min(Index(32), std::max(maxBitsLeft, maxBitsRight) + 1);
+      }
+      case MulInt32: {
+        auto maxBitsRight = getMaxBits(binary->right, localInfoProvider);
+        auto maxBitsLeft = getMaxBits(binary->left, localInfoProvider);
+        return std::min(Index(32), maxBitsLeft + maxBitsRight);
+      }
+      case DivSInt32: {
+        if (auto* c = binary->right->dynCast<Const>()) {
+          auto maxBitsLeft = getMaxBits(binary->left, localInfoProvider);
+          if (maxBitsLeft == 32) {
+            return 32;
+          }
+          if (maxBitsLeft == 0) {
+            return 0;
+          }
+          auto value = c->value.geti32();
+          auto maxBitsRight = 31 - Index(CountLeadingZeroes(value));
```

Comment:
Does this almost duplicate code from like 144? Why is it duplicated, and why is it different?

Reply:
Sorry, could you made link to this line? [line 144](https://github.com/WebAssembly/binaryen/blob/f9240087d7fe575047fb98546b86031d7f9dc45c/src/ir/bits.h#L144) contain something non-similar to that code.

Reply:
It's the line that computes the max bits for a Const,
```cpp
  if (auto* const_ = curr->dynCast<Const>()) {
    switch (curr->type.getBasic()) {
      case Type::i32:
        return 32 - const_->value.countLeadingZeroes().geti32();
```

Reply:
I see. So you suggest change it to:
```ts
auto maxBitsRight = 31 - c->value.countLeadingZeroes().geti32();
```
or
```ts
auto maxBitsRight = getMaxBits(c, localInfoProvider) - 1;
```
?

Reply:
it should be `maxLeftBits - rightBits + 1`. It's easily check on next example:
`maxBits(0x7FFFFFFF / 3)`  ==  `30`

`maxLeftBits = maxBits(0x7FFFFFFF)` is `31`
`rightBits = maxBits(3)` is `2`

`31 - 2` ==> `29`
`31 - 2 + 1` ==> `30`

also it properly handle case when `maxLeftBits == rightBits`